### PR TITLE
[FLINK-15232][table] Message of NoMatchingTableFactoryException should tell users what's wrong

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Currently, all the required properties should exist and match, otherwise, NoMatchingTableFactoryException will be thrown.
We can pick a best candidate to print where is wrong,  print requiredContext and supportedProperties in different formats.
In this way, users can know which property in requiredContext is lack, and which property is not allow in supportedProperties.

## Brief change log

- print candidates when required context properties mismatch. 
- print candidates when No factory supports all properties.
- following factories just print factory class

## Verifying this change

This change is a trivial rework without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by JingsongLi** [comment]
CC: @wuchong @KurtYoung @twalthr @dawidwys 

The error message will be:

```
The match candidates: 
org.apache.flink.streaming.connectors.kafka.KafkaTableSourceSinkFactory
Mismatched properties: 
connector.version=universal <=> 0.10
```

```
The match candidates: 
org.apache.flink.streaming.connectors.kafka.KafkaTableSourceSinkFactory
Missing properties: 
connector.version=universal
```

```
The match candidates: 
org.apache.flink.table.factories.utils.TestTableFormatFactory
Unsupported property keys: 
format.unknown

The following factories have been considered:
org.apache.flink.table.factories.utils.TestTableFormatFactory
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java [hunk 1]
```diff
@@ -255,10 +256,11 @@
 	private static <T extends TableFactory> List<T> filterByContext(
 			Class<T> factoryClass,
 			Map<String, String> properties,
-			List<TableFactory> foundFactories,
 			List<T> classFactories) {
 
-		List<T> matchingFactories = classFactories.stream().filter(factory -> {
+		List<T> matchingFactories = new ArrayList<>();
+		Tuple3<T, Integer, Map<String, Tuple2<String, String>>> bestMatched = null;
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: Could you create a inner class to encapsulate the matched information? It's really hard to guess what does every part mean.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java [hunk 2]
```diff
@@ -269,16 +271,67 @@
...
-			return plainContext.keySet()
-				.stream()
-				.allMatch(e -> properties.containsKey(e) && properties.get(e).equals(plainContext.get(e)));
-		}).collect(Collectors.toList());
+			int matchedSize = 0;
+			Map<String, Tuple2<String, String>> unMatchedProperties = new HashMap<>();
+			for (Map.Entry<String, String> e : plainContext.entrySet()) {
+				if (properties.containsKey(e.getKey()) &&
+						properties.get(e.getKey()).equals(e.getValue())) {
+					matchedSize++;
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: If we have `unMatchedProperties`, we don't need to track `matchedSize` any more.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java [hunk 3]
```diff
@@ -269,16 +271,67 @@
...
+					}
+				});
+
+				if (missing.size() > 0) {
+					builder.append("Missing properties: ");
+					missing.forEach((k, v) -> builder.append("\n").append(k).append("=").append(v));
+				}
+
+				if (mismatched.size() > 0) {
+					builder.append("Mismatched properties: ");
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]:
```suggestion
					builder.append("\nMismatched properties: ");
```
We should have a line feed to separate with the preceding "Missing properties"

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java [hunk 4]
```diff
@@ -269,16 +271,67 @@
...
+				bestMatched.f2.forEach((key, value) -> {
+					if (value.f1 == null) {
+						missing.put(key, value.f0);
+					} else {
+						mismatched.put(key, value);
+					}
+				});
+
+				if (missing.size() > 0) {
+					builder.append("Missing properties: ");
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]:
```suggestion
					builder.append("\nMissing properties: ");
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java [hunk 5]
```diff
@@ -331,41 +384,40 @@
...
+		if (supportedFactories.isEmpty()) {
+			String bestMatchedMessage = null;
+			if (bestMatched != null) {
+				StringBuilder builder = new StringBuilder();
+				builder
+						.append(bestMatched.f0.getClass().getName())
+						.append("\n")
+						.append("Unsupported property keys: ");
+				bestMatched.f1.forEach(k -> builder.append("\n").append(k));
+				bestMatchedMessage = builder.toString();
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: Can be simplified using `String.format` like the original code.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java [hunk 6]
```diff
@@ -269,16 +271,67 @@
...
+
+				if (mismatched.size() > 0) {
+					builder.append("Mismatched properties: ");
+					bestMatched.f2.entrySet().stream().filter(e -> e.getValue().f1 != null).forEach(e ->
+							builder.append("\n")
+									.append(e.getKey())
+									.append("=")
+									.append(e.getValue().f0)
+									.append(" <=> ")
+									.append(e.getValue().f1));
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]:
```
Mismatched properties: 
connector.version=universal <=> 0.10
```

can be improved to:

```
Mismatched properties: 
'connector.version' expects 'universal', but is '0.10'.
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java (hunk 2 above)
Comment (JingsongLi) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: No, they are different size.
For example, there are two factories: A with 100 properties and B with 10 properties.
Maybe we matched 2 for A and 8 for B, we need choose B, but `unMatchedProperties` of A will be more.
Of course we can check value of `unMatchedProperties`, but I don't want to make it complex.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java (hunk 3 above)
Comment (JingsongLi) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: I'd like to keep it together with factory class name.
If there is a line feed, it is a little hard to know it belongs to `match candidates` and this factory.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java (hunk 2 above)
Comment (wuchong) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: The `matchedSize` can be calculated by `userProperties.size() - unMatchedProperties.size()`.

Comment (wuchong) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: The `matchedSize` can be calculated by `requiredProperties.size() - unMatchedProperties.size()`. Is that right?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java (hunk 3 above)
Comment (wuchong) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: If we don't have the `\n` the message will be:

```
The match candidates: 
org.apache.flink.streaming.connectors.kafka.KafkaTableSourceSinkFactory
Missing properties: 
connector.version=universalMismatched properties: 
connector.xxx=yyy <=> zzz
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java (hunk 2 above)
Comment (JingsongLi) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: I got what are mean, we need split `unMatchedProperties` to `missingProperties` and `mismatchedProperties`.

Comment (JingsongLi) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: Unfortunately, we still don't know, you can take a look to my above example.

Comment (JingsongLi) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: You can take a look again after updating the code.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java (hunk 3 above)
Comment (JingsongLi) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java]: Got your point.

**Comment by JingsongLi** [comment]
Thanks @wuchong for your review, updated.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java [hunk 7]
```diff
@@ -32,6 +32,8 @@
 
 	// message that indicates the current matching step
 	private final String message;
+	// message that indicates the best matched factory
+	private final String matchCandidatesMessage;
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java]: Add `@Nullable` annotation to `matchCandidatesMessage`.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java [hunk 8]
```diff
@@ -41,13 +43,15 @@
 
 	public NoMatchingTableFactoryException(
 		String message,
+		String matchCandidatesMessage,
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java]: Add `@Nullable` annotation to `matchCandidatesMessage` parameter.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java [hunk 9]
```diff
@@ -58,24 +62,32 @@ public NoMatchingTableFactoryException(
...
 
 	@Override
 	public String getMessage() {
+		String matchCandidatesString = matchCandidatesMessage == null ?
+			"" :
+			"The match candidates:\n" + matchCandidatesMessage + "\n\n";
 		return String.format(
 			"Could not find a suitable table factory for '%s' in\nthe classpath.\n\n" +
-				"Reason: %s\n\n" +
+				"Reason: %s\n\n" + matchCandidatesString +
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java]: Use String.format pattern. 

```
"Reason: %s\n\n%s"
```

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/factories/TestTableSinkFactory.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_PROPERTY_VERSION;
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_TYPE;
+import static org.apache.flink.table.descriptors.FormatDescriptorValidator.FORMAT_PROPERTY_VERSION;
+import static org.apache.flink.table.descriptors.FormatDescriptorValidator.FORMAT_TYPE;
+
+/**
+ * Test table sink factory.
+ */
+public class TestTableSinkFactory implements TableFactory, TableSinkFactory<Row> {
```
Comment (wuchong) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/factories/TestTableSinkFactory.java]: Can we remove `TestTableSinkFactory` in flink-table-planner now?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java [hunk 11]
```diff
@@ -58,24 +62,32 @@ public NoMatchingTableFactoryException(
...
+			List<TableFactory> factories,
+			Map<String, String> properties) {
+		this(message, matchCandidatesMessage, factoryClass, factories, properties, null);
 	}
 
 	@Override
 	public String getMessage() {
+		String matchCandidatesString = matchCandidatesMessage == null ?
+			"" :
+			"The match candidates:\n" + matchCandidatesMessage + "\n\n";
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java]: `match` -> `matched` ?

File: flink-table/flink-table-common/src/test/resources/log4j-test.properties [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+# Set root logger level to OFF to not flood build logs
+# set manually to INFO for debugging purposes
+log4j.rootLogger=OFF, testlogger
```
Comment (wuchong) [inline flink-table/flink-table-common/src/test/resources/log4j-test.properties]: Are log4j and logback configuration files required?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java (hunk 11 above)
Comment (JingsongLi) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java]: Not matched, just tell user can match.

**Pull request merged**